File: test/IECore/InterpolatedCache.py
##########################################################################
#
#  Copyright (c) 2007-2009, Image Engine Design Inc. All rights reserved.
#
#  Redistribution and use in source and binary forms, with or without
#  modification, are permitted provided that the following conditions are
#  met:
#
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#
#     * Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in the
#       documentation and/or other materials provided with the distribution.
#
#     * Neither the name of Image Engine Design nor the names of any
#       other contributors to this software may be used to endorse or
#       promote products derived from this software without specific prior
#       written permission.
#
#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
#  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
#  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
#  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
#  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
#  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
#  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
#  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
#  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
#  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
#  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
##########################################################################


"""Unit test for InterpolatedCache binding"""
import os, glob
import unittest
import math

from IECore import *


class TestInterpolatedCache(unittest.TestCase):

	pathTemplate = "./test/cache_####.fio"

	def mark( self ) :

		self.__createCacheFile( 250, 0 )
		self.__createCacheFile( 333, 1 )
		self.__createCacheFile( 416, 2 )
		self.__createCacheFile( 499, 3 )
		self.__createCacheFile( 583, 4 )
		self.__createCacheFile( 666, 5 )
		self.__createCacheFile( 749, 6 )

	def __createV3f( self, value ):
		dataWritten = V3fVectorData()
		for i in range(0, 3):
			dataWritten.append( V3f( value, value, value ) )
		return dataWritten

	def __createCacheFile( self, frame, value ):

		fs = FileSequence( self.pathTemplate, EmptyFrameList() )

		cache = AttributeCache( fs.fileNameForFrame( frame ), IndexedIOOpenMode.Write)

		dataWritten = self.__createV3f( value )
		cache.write( "obj1", "v3fVec", dataWritten)
		cache.write( "obj2", "i", IntData( value ) )
		cache.write( "obj2", "d", DoubleData( value ) )
		cache.writeHeader( "testCache", self.__headerCompound( frame ) )

	def __headerCompound( self, frame ):
		return CompoundObject(
			{
				"a": IntData( 1 ),
				"b": StringData( "HEllo!" ),
				"frame": FloatData( frame ),
			}
		)

	def __time( self, frame ):
		return int( frame * 6000. / 24. )

	def __createCache( self ):
		"""Create some cache files do play with..."""

		self.__createCacheFile( self.__time( 0 ), 0 )
		self.__createCacheFile( self.__time( 1 ), 1 )
		self.__createCacheFile( self.__time( 2 ), 2 )
		self.__createCacheFile( self.__time( 3 ), 4 )
		self.__createCacheFile( self.__time( 4 ), 8 )
		self.__createCacheFile( self.__time( 5 ), 16 )

	def testConstructors(self):
		"""Test InterpolatedCache constructors"""
		cache = InterpolatedCache(self.pathTemplate, frame = 0, interpolation = InterpolatedCache.Interpolation.None )
		self.assertEqual( cache.getInterpolation(), InterpolatedCache.Interpolation.None )
		self.assertEqual( cache.getPathTemplate(), self.pathTemplate )
		self.assertEqual( cache.getFrame(), 0 )

		os = OversamplesCalculator( frameRate = 20, samplesPerFrame = 1 )
		cache = InterpolatedCache(self.pathTemplate, frame = 0, interpolation = InterpolatedCache.Interpolation.Linear, oversamplesCalculator = os )
		self.assertEqual( cache.getInterpolation(), InterpolatedCache.Interpolation.Linear )

		os = OversamplesCalculator( frameRate = 24, samplesPerFrame = 1 )
		cache = InterpolatedCache(self.pathTemplate, frame = 0, interpolation = InterpolatedCache.Interpolation.Cubic, oversamplesCalculator = os )

	def testHeaders( self ):
		"""Test InterpolatedCache headers"""

		self.__createCache()

		cache = InterpolatedCache(self.pathTemplate, frame = 1.5, interpolation = InterpolatedCache.Interpolation.Linear )
		self.assert_( "testCache" in cache.headers() )
		h = cache.readHeader( "testCache" )
		self.assertEqual( h, self.__headerCompound( ( self.__time( 1 ) + self.__time( 2 ) ) / 2 ) )

	def testAttributes(self):
		"""Test InterpolatedCache attributes"""

		self.__createCache()

		cache = InterpolatedCache(self.pathTemplate, frame = 1.2, interpolation = InterpolatedCache.Interpolation.Linear )
		self.assertEqual( len( cache.attributes( "obj2" )), 2 )
		self.assertEqual( len( cache.attributes( "obj2", "i.*" )), 1 )
		self.assertEqual( len( cache.attributes( "obj1" )), 1 )

	def testContains(self):
		"""Test InterpolatedCache contains"""

		self.__createCache()

		cache = InterpolatedCache(self.pathTemplate, frame = 1.2, interpolation = InterpolatedCache.Interpolation.Linear )
		self.assert_( cache.contains( "obj1" ) )
		self.assert_( cache.contains( "obj2", "i" ) )
		self.failIf( cache.contains( "obj3" ) )
		self.failIf( cache.contains( "obj2", "not_in_cache" ) )

	def testReading(self):
		"""Test InterpolatedCache read"""

		self.__createCache()

		cache = InterpolatedCache(self.pathTemplate, frame = 1.5, interpolation = InterpolatedCache.Interpolation.Linear )
		self.assertEqual( cache.read( "obj2", "i" ), IntData( 1 ) )
		self.assertEqual( cache.read( "obj2", "d" ), DoubleData( 1.5 ) )
		self.assertEqual( cache.read( "obj1" ), CompoundObject( { "v3fVec": self.__createV3f( 1.5 ) } ) )

	def testOversampledReading( self ) :

		self.mark()

		os = OversamplesCalculator( frameRate = 24, samplesPerFrame = 3 )
		cache = InterpolatedCache(self.pathTemplate, frame = 1, interpolation = InterpolatedCache.Interpolation.Linear, oversamplesCalculator = os )
		cache.setFrame( 2.0 )
		self.assertAlmostEqual( cache.read( "obj2", "d" ).value, 3, 4 )


	def testChangingParameters( self ):
		"""Test InterpolatedCache reading and changing frame, interpolation, ... """

		self.__createCache()

		cache = InterpolatedCache(self.pathTemplate, frame = 0, interpolation = InterpolatedCache.Interpolation.None )
		self.assertEqual( cache.read( "obj2", "d" ), DoubleData( 0 ) )
		cache.setFrame( 0.8 )
		self.assertEqual( cache.read( "obj2", "d" ), DoubleData( 0 ) )
		cache.setFrame( 1 )
		self.assertEqual( cache.read( "obj2", "d" ), DoubleData( 1 ) )
		cache.setFrame( 0.5 )
		cache.setInterpolation( InterpolatedCache.Interpolation.Linear )
		self.assertEqual( cache.read( "obj2", "d" ), DoubleData( 0.5 ) )
		cache.setFrame( 2.25 )
		self.assertAlmostEqual( cache.read( "obj2", "d" ).value, 2.5, 3 )
		cache.setInterpolation( InterpolatedCache.Interpolation.Cubic )
		self.assertAlmostEqual( cache.read( "obj2", "d" ).value, 2.4531, 3 )
		cache.setFrame( 5 )
		self.assertEqual( cache.read( "obj2", "d" ), DoubleData( 16 ) )
		cache.setInterpolation( InterpolatedCache.Interpolation.Linear )
		self.assertEqual( cache.read( "obj2", "d" ), DoubleData( 16 ) )
		cache.setInterpolation( InterpolatedCache.Interpolation.None )
		self.assertEqual( cache.read( "obj2", "d" ), DoubleData( 16 ) )

	def testOldTransformationMatrixData( self ):
		cache = InterpolatedCache( "test/IECore/data/attributeCaches/transform.old.####.fio", frame = 4, interpolation = InterpolatedCache.Interpolation.Linear )
		self.assertEqual( cache.read( ".parent", "transformCache.transform" ).value.rotate, Eulerd() )
		self.assertAlmostEqual( (cache.read( ".pSphere1", "transformCache.transform" ).value.rotate - Eulerd( 0.244525, 0, 0 )).length(), 0, 2 )
		cache.setFrame( 4.5 )
		self.assertEqual( cache.read( ".parent", "transformCache.transform" ).value.rotate, Eulerd() )
		self.assertAlmostEqual( (cache.read( ".pSphere1", "transformCache.transform" ).value.rotate - Eulerd(0.283422, 0, 0)).length(), 0, 2 )

	def testNewTransformationMatrixData( self ):

		# Code that created the cache files for this test:
		#cache1 = AttributeCache( "test/IECore/data/attributeCaches/transform.new.0250.fio", IndexedIOOpenMode.Write )
		#cache2 = AttributeCache( "test/IECore/data/attributeCaches/transform.new.0500.fio", IndexedIOOpenMode.Write )
		#for order in [ Eulerd.Order.XYZ, Eulerd.Order.XZY, Eulerd.Order.YZX, Eulerd.Order.YXZ, Eulerd.Order.ZXY, Eulerd.Order.ZYX ]:
		#	t = TransformationMatrixd()
		#	t.rotate = Eulerd( 1000, -10, 100, order )
		#	transform = TransformationMatrixdData( t )
		#	cache1.write( "test%d" % order, "transform", transform )
		#	t = TransformationMatrixd()
		#	t.rotate = Eulerd( 1002, -12, 102, order )
		#	transform = TransformationMatrixdData( t )
		#	cache2.write( "test%d" % order, "transform", transform )
		#cache1 = None
		#cache2 = None

		cache = InterpolatedCache( "test/IECore/data/attributeCaches/transform.new.####.fio", frame = 1.5, interpolation = InterpolatedCache.Interpolation.Linear )
		for order in [ Eulerd.Order.XYZ, Eulerd.Order.XZY, Eulerd.Order.YZX, Eulerd.Order.YXZ, Eulerd.Order.ZXY, Eulerd.Order.ZYX ]:
			self.assertEqual( cache.read( "test%d" % order, "transform" ).value.rotate, Eulerd( 1001, -11, 101, order ) )
		cache = None

	def tearDown(self):

		# cleanup
		allFiles = glob.glob( self.pathTemplate.replace( "####", "*" ) )
		for f in allFiles:
			os.remove( f )

if __name__ == "__main__":
	unittest.main()

File: contrib/removedInCortex9/attributeCache/src/IECore/InterpolatedCache.cpp
//////////////////////////////////////////////////////////////////////////
//
//  Copyright (c) 2007-2011, Image Engine Design Inc. All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are
//  met:
//
//     * Redistributions of source code must retain the above copyright
//       notice, this list of conditions and the following disclaimer.
//
//     * Redistributions in binary form must reproduce the above copyright
//       notice, this list of conditions and the following disclaimer in the
//       documentation and/or other materials provided with the distribution.
//
//     * Neither the name of Image Engine Design nor the names of any
//       other contributors to this software may be used to endorse or
//       promote products derived from this software without specific prior
//       written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
//  IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
//  THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
//  PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR
//  CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
//  EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
//  PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
//  PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
//
//////////////////////////////////////////////////////////////////////////

#include <cassert>

#include "tbb/mutex.h"

#include "boost/format.hpp"
#include "boost/bind.hpp"

#include "IECore/MessageHandler.h"
#include "IECore/OversamplesCalculator.h"
#include "IECore/ObjectInterpolator.h"
#include "IECore/InterpolatedCache.h"
#include "IECore/CompoundObject.h"
#include "IECore/FileSequence.h"
#include "IECore/EmptyFrameList.h"
#include "IECore/LRUCache.h"

using namespace IECore;
using namespace boost;

//////////////////////////////////////////////////////////////////////////
// Private implementation class
//////////////////////////////////////////////////////////////////////////

class InterpolatedCache::Implementation : public IECore::RefCounted
{

	public :
	
		Implementation( const std::string &pathTemplate, Interpolation interpolation, const OversamplesCalculator &o, size_t maxOpenFiles )
			:	m_cachesForTicks( bind( &Implementation::cachesForTicksGetter, this, _1, _2 ), maxOpenFiles )
		{
			if( pathTemplate.size() )
			{
				setPathTemplate( pathTemplate );
			}
			setInterpolation( interpolation );
			setOversamplesCalculator( o );
		}
	
		void setPathTemplate( const std::string &pathTemplate )
		{
			if ( !m_fileSequence || getPathTemplate() != pathTemplate )
			{
				m_fileSequence = new FileSequence( pathTemplate, new EmptyFrameList() );
				m_cachesForTicks.clear();
			}
		}

		const std::string &getPathTemplate() const
		{
			if( !m_fileSequence )
			{
				// i'd prefer to return "" but that would mean returning a reference
				// to a temporary or having some weird static string instance.
				throw Exception( "Path template has not been set" );
			}
			return m_fileSequence->getFileName();
		}
		
		void setMaxOpenFiles( size_t maxOpenFiles )
		{
			m_cachesForTicks.setMaxCost( maxOpenFiles );
		}
		
		size_t getMaxOpenFiles() const
		{
			return m_cachesForTicks.getMaxCost();
		}
	
		void setInterpolation( Interpolation interpolation )
		{
			m_interpolation = interpolation;
		}

		Interpolation getInterpolation() const
		{
			return m_interpolation;
		}

		void setOversamplesCalculator( const OversamplesCalculator &oc )
		{
			m_oversamplesCalculator = oc;
		}

		const OversamplesCalculator &getOversamplesCalculator() const
		{
			return m_oversamplesCalculator;
		}

		ObjectPtr read( float frame, const ObjectHandle &obj, const AttributeHandle &attr ) const
		{
			CacheAndMutexPtr c[4]; float x = 0;
			int numCaches = caches( frame, c, x );
			
			assert( numCaches );
			
			ObjectPtr r[4];
			for( int i=0; i<numCaches; i++ )
			{
				/// \todo Can we launch each of these reads in a separate thread?
				CacheAndMutex::Mutex::scoped_lock lock( c[i]->mutex );
				r[i] = c[i]->cache->read( obj, attr );
			}
			
			ObjectPtr result = 0;
			if( numCaches > 1 )
			{
				switch( m_interpolation )
				{
					case Linear :
						assert( numCaches==2 );
						result = linearObjectInterpolation( r[0].get(), r[1].get(), x );
						break;
					case Cubic :
						assert( numCaches==4 );
						result = cubicObjectInterpolation( r[0].get(), r[1].get(), r[2].get(), r[3].get(), x );
						break;
					default :
						assert( false );
				}
			}
			
			if( !result )
			{
				// either there was only one cache, or interpolation failed.
				// in both cases we just return the first sample.
				result = r[0];
			}
			
			assert( result );
			return result;
		}
		
		CompoundObjectPtr read( float frame, const ObjectHandle &obj ) const
		{
			CompoundObjectPtr result = new CompoundObject();

			std::vector<AttributeHandle> attrs;
			attributes( frame, obj, attrs );

			for( std::vector<AttributeHandle>::const_iterator it = attrs.begin(); it != attrs.end(); ++it )
			{
				/// \todo currently each function call is recalculating a bunch of stuff about which caches
				/// to use and wotnot. this could be done once and shared.
				result->members()[ *it ] = read( frame, obj, *it );
			}

			return result;
		}

		ObjectPtr readHeader( float frame, const HeaderHandle &hdr ) const
		{
			CacheAndMutexPtr c[4]; float x = 0;
			int numCaches = caches( frame, c, x );
			
			ObjectPtr r[4];
			for( int i=0; i<numCaches; i++ )
			{
				/// \todo Can we launch each of these reads in a separate thread?
				CacheAndMutex::Mutex::scoped_lock lock( c[i]->mutex );
				r[i] = c[i]->cache->readHeader( hdr );
			}
			
			ObjectPtr result = 0;
			if( numCaches > 1 )
			{
				switch( m_interpolation )
				{
					case Linear :
						assert( numCaches==2 );
						result = linearObjectInterpolation( r[0].get(), r[1].get(), x );
						break;
					case Cubic :
						assert( numCaches==4 );
						result = cubicObjectInterpolation( r[0].get(), r[1].get(), r[2].get(), r[3].get(), x );
						break;
					default :
						assert( false );
				}
			}
			
			if( !result )
			{
				// either there was only one cache, or interpolation failed.
				// in both cases we just return the first sample.
				result = r[0];
			}
			
			assert( result );
			return result;
		}

		CompoundObjectPtr readHeader( float frame ) const
		{
			CompoundObjectPtr result = new CompoundObject();
			
			std::vector<HeaderHandle> hds;
			headers( frame, hds );

			for( std::vector<HeaderHandle>::const_iterator it = hds.begin(); it != hds.end(); ++it )
			{
				/// \todo currently each function call is recalculating a bunch of stuff about which caches
				/// to use and wotnot. this could be done once and shared.
				result->members()[ *it ] = readHeader( frame, *it );

			}
			return result;
		}

		void objects( float frame, std::vector<ObjectHandle> &objs ) const
		{
			CacheAndMutexPtr c[4]; float x = 0;
			int numCaches = caches( frame, c, x );
			assert( numCaches ); (void)numCaches;
			CacheAndMutex::Mutex::scoped_lock lock( c[0]->mutex );
			c[0]->cache->objects( objs );
		}

		void attributes( float frame, const ObjectHandle &obj, std::vector<AttributeHandle> &attrs ) const
		{
			CacheAndMutexPtr c[4]; float x = 0;
			int numCaches = caches( frame, c, x );
			assert( numCaches ); (void)numCaches;
			CacheAndMutex::Mutex::scoped_lock lock( c[0]->mutex );
			c[0]->cache->attributes( obj, attrs );
		}

		void attributes( float frame, const ObjectHandle &obj, const std::string regex, std::vector<AttributeHandle> &attrs ) const
		{
			CacheAndMutexPtr c[4]; float x = 0;
			int numCaches = caches( frame, c, x );
			assert( numCaches ); (void)numCaches;
			CacheAndMutex::Mutex::scoped_lock lock( c[0]->mutex );
			c[0]->cache->attributes( obj, regex, attrs );
		}

		void headers( float frame, std::vector<HeaderHandle> &hds ) const
		{
			CacheAndMutexPtr c[4]; float x = 0;
			int numCaches = caches( frame, c, x );
			assert( numCaches ); (void)numCaches;
			CacheAndMutex::Mutex::scoped_lock lock( c[0]->mutex );
			c[0]->cache->headers( hds );
		}
		
		bool contains( float frame, const ObjectHandle &obj ) const
		{
			CacheAndMutexPtr c[4]; float x = 0;
			int numCaches = caches( frame, c, x );
			assert( numCaches ); (void)numCaches;
			CacheAndMutex::Mutex::scoped_lock lock( c[0]->mutex );
			return c[0]->cache->contains( obj );
		}
		
		bool contains( float frame, const ObjectHandle &obj, const AttributeHandle &attr ) const
		{
			CacheAndMutexPtr c[4]; float x = 0;
			int numCaches = caches( frame, c, x );
			assert( numCaches ); (void)numCaches;
			CacheAndMutex::Mutex::scoped_lock lock( c[0]->mutex );
			return c[0]->cache->contains( obj, attr );
		}

	private :
	
		FileSequencePtr m_fileSequence;
		Interpolation m_interpolation;
		OversamplesCalculator m_oversamplesCalculator;
			
		// a class to store an AttributeCache and a mutex
		// that must be used to access it.
		class CacheAndMutex : public IECore::RefCounted
		{
			public :

				typedef tbb::mutex Mutex;
				Mutex mutex;	
				AttributeCachePtr cache;

		};
		
		IE_CORE_DECLAREPTR( CacheAndMutex );

		// an lru cache mapping from ticks to attribute caches

		CacheAndMutexPtr cachesForTicksGetter( const int &tick, size_t &cost )
		{			
			if( !m_fileSequence )
			{
				throw Exception( "Path template has not been set" );
			}
			
			std::string fileName = m_fileSequence->fileNameForFrame( tick );
			CacheAndMutexPtr result = new CacheAndMutex;
			result->cache = new AttributeCache( fileName, IndexedIO::Read );
			return result;
		}
		
		typedef LRUCache<int, CacheAndMutexPtr> CachesForTicks;
		mutable CachesForTicks m_cachesForTicks;
		
		// function to find the relevant caches for a given frame and return
		// them along with an interpolation type and factor. returns the number
		// of caches found. note that this may be 1 even if interpolation has
		// been requested, as the frame may coincide directly with a file cache
		// and not need interpolating.
		
		int caches( float frame, CacheAndMutexPtr c[4], float &interpolationFactor ) const
		{
		
			int lowTick, highTick;
			interpolationFactor = m_oversamplesCalculator.tickInterval( frame, lowTick, highTick );

			float step = (float)m_oversamplesCalculator.getTicksPerSecond() / m_oversamplesCalculator.getSamplesPerFrame() / m_oversamplesCalculator.getFrameRate();

			int start = 0;
			int end = 0;
			if( interpolationFactor == 0.0f || m_interpolation == None )
			{
				start = 0;
				end = 0;
			}
			else
			{
				switch( m_interpolation )
				{
					case Linear:
						// Need one file ahead
						start = 0;
						end = 1;
						break;
					case Cubic:
						// Need one file behind, 2 files ahead
						start = -1;
						end = 2;
						break;
					default:
						assert( false );
				}
			}

			int cacheIndex = 0;
			for( int fileNum = start; fileNum <= end; fileNum++, cacheIndex++ )
			{				
				int tick = m_oversamplesCalculator.nearestTick(( int )( lowTick + fileNum * step ) );				
				c[cacheIndex] = m_cachesForTicks.get( tick );
			}

			assert( cacheIndex );
			return cacheIndex;
		}
	

};

//////////////////////////////////////////////////////////////////////////
// InterpolatedCache class
//////////////////////////////////////////////////////////////////////////

InterpolatedCache::InterpolatedCache( const std::string &pathTemplate, Interpolation interpolation,  const OversamplesCalculator &o, size_t maxOpenFiles )
	:	m_implementation( new Implementation( pathTemplate, interpolation, o, maxOpenFiles ) )
{
}

InterpolatedCache::~InterpolatedCache()
{
}

void InterpolatedCache::setPathTemplate( const std::string &pathTemplate )
{
	m_implementation->setPathTemplate( pathTemplate );
}

const std::string &InterpolatedCache::getPathTemplate() const
{
	return m_implementation->getPathTemplate();
}

void InterpolatedCache::setMaxOpenFiles( size_t maxOpenFiles )
{
	m_implementation->setMaxOpenFiles( maxOpenFiles );
}

size_t InterpolatedCache::getMaxOpenFiles() const
{
	return m_implementation->getMaxOpenFiles();
}

void InterpolatedCache::setInterpolation( InterpolatedCache::Interpolation interpolation )
{
	m_implementation->setInterpolation( interpolation );
}

InterpolatedCache::Interpolation InterpolatedCache::getInterpolation() const
{
	return m_implementation->getInterpolation();
}

void InterpolatedCache::setOversamplesCalculator( const OversamplesCalculator &oc )
{
	m_implementation->setOversamplesCalculator( oc );
}

const OversamplesCalculator &InterpolatedCache::getOversamplesCalculator() const
{
	return m_implementation->getOversamplesCalculator();
}

ObjectPtr InterpolatedCache::read( float frame, const ObjectHandle &obj, const AttributeHandle &attr ) const
{
	return m_implementation->read( frame, obj, attr );
}

CompoundObjectPtr InterpolatedCache::read( float frame, const ObjectHandle &obj ) const
{
	return m_implementation->read( frame, obj );
}

ObjectPtr InterpolatedCache::readHeader( float frame, const HeaderHandle &hdr ) const
{
	return m_implementation->readHeader( frame, hdr );
}

CompoundObjectPtr InterpolatedCache::readHeader( float frame ) const
{
	return m_implementation->readHeader( frame );
}

void InterpolatedCache::objects( float frame, std::vector<ObjectHandle> &objs ) const
{
	return m_implementation->objects( frame, objs );
}

void InterpolatedCache::headers( float frame, std::vector<HeaderHandle> &hds ) const
{
	return m_implementation->headers( frame, hds );
}

void InterpolatedCache::attributes( float frame, const ObjectHandle &obj, std::vector<AttributeHandle> &attrs ) const
{
	return m_implementation->attributes( frame, obj, attrs );
}

void InterpolatedCache::attributes( float frame, const ObjectHandle &obj, const std::string regex, std::vector<AttributeHandle> &attrs ) const
{
	return m_implementation->attributes( frame, obj, regex, attrs );
}

bool InterpolatedCache::contains( float frame, const ObjectHandle &obj ) const
{
	return m_implementation->contains( frame, obj );
}

bool InterpolatedCache::contains( float frame, const ObjectHandle &obj, const AttributeHandle &attr ) const
{
	return m_implementation->contains( frame, obj, attr );
}
